Make the compiler resilient to broken IDE refactorings

gafter opened the pull request:
This makes the compiler resilient to bad syntax trees that result from broken
IDE refactorings, so the compiler will not crash and take the IDE with it.
Related to #17321 

@dotnet/roslyn-compiler Please review.
@AlekseyTs You may be interested to see that this uses the distinction with `LocalDeclarationKind.OutVar`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -2213,7 +2213,20 @@ private BoundExpression BindOutVariableArgument(DeclarationExpressionSyntax decl
                 if (isVar)
                 {
-                    return new OutVariablePendingInference(declarationExpression, localSymbol, null);
+                    if (localSymbol.DeclarationKind == LocalDeclarationKind.OutVariable)
```
>if (localSymbol.DeclarationKind == LocalDeclarationKind.OutVariable) [](start = 20, length = 68)

I believe that this method should never be called if this condition is not true. The fix should be applied elsewhere.

AlekseyTs replied:
BindArgumentValue should never call this method if the argument is not an Out Var Declaration.

---
In reply to: [103062578](https://github.com/dotnet/roslyn/pull/17380#discussion_r103062578) [](ancestors = 103062578)

gafter commented:
@AlekseyTs I've made changes as suggested in your previous review.
@dotnet/roslyn-compiler May I have a second review, please?

jcouv commented on src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs:
```diff
@@ -377,6 +377,12 @@ internal static CSharpSyntaxNode GetContainingDeconstruction(this ExpressionSynt
+        {
+            return p.Parent?.Kind() == SyntaxKind.Argument
+                && ((ArgumentSyntax)p.Parent).RefOrOutKeyword.Kind() == SyntaxKind.OutKeyword;
+        }
```
I didn't understand why this method is needed. Why not use the one below?

AlekseyTs replied:
I guess, the existing method doesn't cover a discard scenario.

cston replied:
Consider implementing the existing method `IsOutVarDeclaration` using the new method `IsOutDeclaration`.
```
return p.Designation.Kind() == SyntaxKind.SingleVariableDesignation && p.IsOutDeclaration();
```

gafter replied:
@cston Done.

jcouv replied:
Thanks

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -30194,6 +30198,105 @@ public static int TakeOutParam<T>(out T x)
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe.WithScriptClassName("Script"), parseOptions: TestOptions.Script);
+
+            compilation.GetDeclarationDiagnostics().Verify(
```
>compilation.GetDeclarationDiagnostics().Verify( [](start = 12, length = 47)

I would expect to have some parse errors. Could we assert them too?

gafter replied:
I'll assert the compilation diagnostics. I don't intend to assert the division of diagnostics between the parser and semantic analysis here.

gafter replied:
By the way, with the current design I would not expect any syntax errors.

AlekseyTs replied:
> By the way, with the current design I would not expect any syntax errors.

Could you elaborate what is "the current design" in this context.
Does it mean that ```H.M((int x1, int x2));``` would have no errors? Or there would be binding errors?

gafter replied:
I would expect binding errors for `H.M((int x1, int x2));` but not syntax errors.

AlekseyTs replied:
I think that is fine. I would be useful to assert them though.

gafter replied:
I am adding code to assert all errors. However, note that this test is for a scenario that is not intended to be fixed in this PR, and remains skipped.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -1028,7 +1028,11 @@ private static void VerifyModelForOutVarInNotExecutableCode(SemanticModel model,
             var typeInfo = model.GetTypeInfo(decl);
             Assert.Equal(expectedType, typeInfo.Type);
-            Assert.Equal(expectedType, typeInfo.ConvertedType);
+            if ((expectedSymbol as LocalSymbol)?.DeclarationKind == LocalDeclarationKind.OutVariable)
```
>if ((expectedSymbol as LocalSymbol)?.DeclarationKind == LocalDeclarationKind.OutVariable) [](start = 12, length = 89)

It looks like this method is also called for fields, the condition is not covering them. Also, it is not clear why the assert for ```ConvertedType``` fails for misplaced declaration expressions. It doesn't look like it fails for the ```Type``` property. It feels like we should assert something for scenarios like this (is it null, is it an error type, is it ...) rather than do no validation because ```ConvertedType``` is a public API.

gafter replied:
I'll change this so it handles fields too. However, I'm not intending to address https://github.com/dotnet/roslyn/issues/17377 (regarding misplaced declaration expressions) in the context of this PR.

AlekseyTs commented:
Done with review pass.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -30194,6 +30198,105 @@ public static int TakeOutParam<T>(out T x)
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe.WithScriptClassName("Script"), parseOptions: TestOptions.Script);
+
+            compilation.GetDeclarationDiagnostics().Verify(
+                // (2,10): error CS7019: Type of 'x1' cannot be inferred since its initializer directly or indirectly refers to the definition.
```
>// (2,10): error CS7019: Type of 'x1' cannot be inferred since its initializer directly or indirectly refers to the definition. [](start = 16, length = 127)

This error is very misleading, the statement that declares the variable doesn't have any direct or indirect references to it.

gafter replied:
This test is skipped pending some future fix for https://github.com/dotnet/roslyn/issues/17377. I'll let whoever works on that consider how much effort to invest in reducing cascaded diagnostics for this unlikely scenario.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -1028,7 +1028,11 @@ private static void VerifyModelForOutVarInNotExecutableCode(SemanticModel model,
-            Assert.Equal(expectedType, typeInfo.ConvertedType);
+            if (decl.IsOutDeclaration())
+            {
+                Assert.Equal(expectedType, typeInfo.ConvertedType);
```
>typeInfo.ConvertedType [](start = 43, length = 22)

What do we expect this to be when the ```if``` condition is not satisfied?

gafter replied:
It is just an expression with a type, so it can be converted to whatever type is required in its context. The infrastructure for these tests do not support testing for that, and it isn't relevant to the bug being fixed (a crash).

AlekseyTs replied:
It looks like some of the existing tests that take advantage of this helper do not satisfy the ```decl.IsOutDeclaration()``` condition and, therefore, would lose test coverage due to this change. Could you please adjust the condition so that it only affects the scenarios where declaration expression can be converted to a different type?

AlekseyTs replied:
I am also not comfortable with simply not testing behavior. There are some expectations for each scenario, when they do not match the ```expectedType```, we could pass expectations via dedicated parameter and verify against that.  Then the syntax check wouldn't be needed at all.

gafter commented:
@dotnet/roslyn-compiler May I please have a couple of reviews for this bug fix?

gafter commented:
@AlekseyTs As you requested, I have restored the incorrect assertion, added comments explaining that, and modified the tests to avoid demonstrating its failure. Do you have any further issues?

@dotnet/roslyn-compiler May I please have a review of this crash fix?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -30197,6 +30208,119 @@ public static int TakeOutParam<T>(out T x)
+                Diagnostic(ErrorCode.ERR_DeclarationExpressionNotPermitted, "var x1").WithLocation(4, 20)
+                );
+
+            // We do not currently have test helpers that support declaration expressions subject to a conversion.
```
>We do not currently have test helpers that support declaration expressions subject to a conversion. [](start = 15, length = 99)

I don't think this should stop us from testing completely. There are plenty of other SemanticModel APIs that could be tested and are not subject for [https://github.com/dotnet/roslyn/issues/17463](https://github.com/dotnet/roslyn/issues/17463).

gafter replied:
@AlekseyTs It is not the intent of this PR to set a new policy of exhaustively exercising the APIs of the compiler in the face of syntax trees that could not have been built by the parser. If you think that is a valuable exercise, please file an issue explaining your justification. I do not intend to take that direction in this PR. This PR intends to work around one particular path in the compiler that causes a crash in a scenario where the IDE builds a bad node.

AlekseyTs replied:
And, I believe, we should test that IDE will get the right information from SemanticModel APIs in such cases, since it can get in them. 

> We do not currently have test helpers

If existing helpers are insufficient, it is very easy to create the ones that would fit the need.

gafter replied:
There is no concept of "right information" for parse trees that the parser cannot construct. I am only trying to remove an IDE crash.

AlekseyTs replied:
> There is no concept of "right information" for parse trees that the parser cannot construct.

I think there is such a concept. We should call the APIs, see if they crash, see what do they return, does the result feel right. If it doesn't feel right, we could decide to follow-up on that later. In my opinion there is definitely a value in capturing the current behavior of the APIs, so that we can catch any changes in behavior and avoid regressions in the future.

gafter replied:
The work you propose is outside the scope of the intended fix here.

AlekseyTs replied:
> The work you propose is outside the scope of the intended fix here.

I am not suggesting to change product behavior, i.e. the fix. I am suggesting that it is not adequately covered by the tests. How do we know if the fix is complete if you don't test it? It would take less than 5 minutes to parameterize existing helpers to create a specialized flavor that would handle your special case, or possibly to adjust the scenario so that it would be handled by the existing helpers. It would probably take less than 15 minutes to create a new brand new dedicated helper from scratch. Yet we are discussing this for more than 4 hours. I am not even asking to test new scenarios, just asking to do more thorough testing for the ones you already have.

cston replied:
We should call `model.GetSymbolInfo(declaration)`, `model.GetTypeInfo(declaration)`, etc. to verify the methods complete successfully, now and in the future.

gafter commented:
@cston @jcouv If you have no further issues with this, can you please approve?

cston commented:
LGTM
